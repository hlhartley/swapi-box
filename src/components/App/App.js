import React, { Component } from 'react';
import '../../main.scss';
import Header from '../Header/Header';
import Navigation from '../Navigation/Navigation';
import ScrollText from '../ScrollText/ScrollText';
import CardContainer from '../CardContainer/CardContainer';

class App extends Component {
  constructor() {
    super();
    this.state = {
      navFixed: false,
      initialNavPosition: 0,
      people: [],
      planets: [],
      vehicles: [],
      selected: ''
    };
  }

  async componentDidMount() {
    await this.fetchPeople()
    this.fetchPlanets()
    this.fetchVehicles()
    window.addEventListener('scroll', this.fixNav.bind(this))
    this.setState({initialNavPosition: 250})
  }

  fixNav() {
    if(window.scrollY >= this.state.initialNavPosition) {
        this.setState({navFixed: true})
    } else {
        this.setState({navFixed: false})
    }
  }

  // TAKE THIS AWAY IF YOU WANT (decorating each person object with an ID could save you multiple fetch calls later on)
  // For example, when you are trying to get a planet's residents you could:

  // 1. Use the planet.residents array of URLs like an array of IDs with the same function below
  // 2. Now you can lookup those people from the list of people you already have from this fetchPeople call (instead of making fetch calls for each for each residentURL)
  // 3. You could speed up this lookup even more by generating a people 'hash' or object instead of a people array:

      // peopleHash = {
      //   1: { Person object} // or whatever data you need to display, maybe just the person's name?
      //   2: { Person object}
      //   3: { Person object}
      // }

      // SIDE NOTE: when you look at efficiency of algorithms you want to consider TIME and SPACE (memory usage)
      // Looking things up in a hash/object is much faster than searching through an array, but by making this hash of {IDs => people objects},
      // we do have to use more memory on the client's computer.

  // 4. You could then get the planet's residents with something like planet.residents.map(resident => peopleHash[this.extractID(resident)])


  extractID(url) {
    // Probably a more efficient way to do this... Perhaps some regex would be faster?
    return url.split('/').slice(-2, -1)[0];
  }

  async fetchPeople() {
    let peopleArray = []
    // for(let i=1; i < 10; i++) {
      const url = `https://swapi.co/api/people/`
      const response = await fetch(url)
      const people = await response.json()
      const peopleWithIds = people.results.map(person => {
        person.id = this.extractID(person.url);
        return person;
      })
      const peopleWithHometowns = await this.fetchHomeworld(peopleWithIds)
      // const peopleWithHometowns = await this.fetchHomeworld(people.results)
      const peopleWithSpecies = await this.fetchSpecies(peopleWithHometowns)
      peopleArray.push(...peopleWithSpecies)
    // }
    this.setState({people: peopleArray})
  }

  fetchHomeworld(people) {
    // return Promise.all(people.map(async (member) => {
    //   const response = await fetch(member.homeworld)
    //   const homeworld = await response.json()
    //   return { 
    //     name: member.name, 
    //     homeworld: homeworld.name, 
    //     population: homeworld.population 
    //   }
    // }))
    const unresolvedPromises = people.map((member) => {
      return fetch(member.homeworld)
      .then(response => response.json())
      .then(homeworld => ({ id: member.id, name: member.name, homeworld: homeworld.name, population: homeworld.population, species: member.species[0] }))
    })
    return Promise.all(unresolvedPromises)
  }

  fetchSpecies(people) {
    const unresolvedPromises = people.map((member) => {
      if(member.species) {
        return fetch(member.species)
        .then(response => response.json())
        .then(species => ({ id: member.id, name: member.name, homeworld: member.homeworld, population: member.population, species: species.name, language: species.language }))
      } else {
        return ({ id: member.id, name: member.name, homeworld: member.homeworld, population: member.population, species: 'N/A', language: 'N/A' })
      }
    })
    return Promise.all(unresolvedPromises)
    // return Promise.all(people.map(async (member) => {
    //   const response = await fetch(member.species[0])
    //   const species = await response.json()
    //   return ({species: species.name})
    // }))
  }

  fetchPlanets() {
    fetch('https://swapi.co/api/planets/')
      .then(response => response.json())
      .then(result => this.setState({planets: result.results}))
      .catch(error => console.log(error))
  }

  fetchVehicles() {
    fetch('https://swapi.co/api/vehicles/')
      .then(response => response.json())
      .then(result => this.setState({vehicles: result.results}))
      .catch(error => console.log(error))
  }

  receiveSelected = (selectedButton) => {
    this.setState({ selected: selectedButton })
  }

  returnCards = () => {
      return this.state[this.state.selected] || []          
  }

  renderScreen = () => {
    if(this.state.selected === '') {
      return <ScrollText />
    } else {
      return <CardContainer category={this.returnCards()}/>
    }
  }
  
  render() {
    return (
      <div className="App">
        <Header />
        <Navigation 
          navFixed={this.state.navFixed}
          iitialNavPosition={this.state.initialNavPosition}
          people={this.state.people}
          receiveSelected={this.receiveSelected}
          // fetchPeople={this.fetchPeople}
          // fetchPlanets={this.fetchPlanets}
          // fetchVehicles={this.fetchVehicles}
        />
        {/* <h2 className='card-container-title'>{this.state.selected.toUpperCase()}</h2> */}
        {this.renderScreen()}
      </div>
    );
  }
}

export default App;
