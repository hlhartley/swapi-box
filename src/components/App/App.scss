.card-container-title {
    display: flex;
    justify-content: center;
    font-size: 40px;
    margin: 0;
    padding-left: 15px;
    padding-top: 15px;
}