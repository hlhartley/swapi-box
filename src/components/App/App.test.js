import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';

let wrapper;
describe('App', () => {

  beforeEach(() => {
    wrapper = shallow(<App/>); 
  })

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('should exist', () => {   
    expect(wrapper).toBeDefined()
  })

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should render the Header, Navigation, and ScrollText and CardContainer components', () => {
    expect(wrapper.find('Header').length).toEqual(1)
    expect(wrapper.find('Navigation').length).toEqual(1)
    expect(wrapper.find('ScrollText').length).toEqual(1)

    wrapper.instance().renderScreen()
    wrapper.setState({selected: 'people'})
    expect(wrapper.find('CardContainer').length).toEqual(1)
  })

  it('should have a proper default state', () => {
    expect(wrapper.state()).toEqual({
      navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'films',
      favorites: [],
    })  
  })

  it('fixNav method should setState of navFixed based on window.scrollY position', () => {
    const expectedState = {
      navFixed: true,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'films',
      favorites: [],
    };
    
    expect(wrapper.state()).toEqual({navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'films',
      favorites: [],});
    wrapper.instance().fixNav()
    expect(wrapper.state()).toEqual(expectedState);
  })

  it('extractIdFrom method should extract the people id from the URL', () => {
    const mockUrl = 'https://swapi.co/api/people/5/'
    const peopleId = wrapper.instance().extractIdFrom(mockUrl)
    expect(peopleId).toEqual('5')

    const mockUrl2 = 'https://swapi.co/api/people/37/'
    const peopleId2 = wrapper.instance().extractIdFrom(mockUrl2)
    expect(peopleId2).toEqual('37')
  })

  it('receiveSelected method should set state of selected to selectedButton', () => {
    wrapper.instance().receiveSelected('people')
    expect(wrapper.state()).toEqual({navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'people',
      favorites: []});
  })

  it('returnCards method returns array of all objects in state of a certain category', () => {
    wrapper.setState({ 
      navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: ['a', 'b', 'c'],
      planets: [1, 2, 3],
      vehicles: ['hi', 'bye'],
      selected: 'planets',
      favorites: [],
    });
    expect(wrapper.instance().returnCards()).toEqual([1, 2, 3]);

    wrapper.setState({ 
      selected: '',
    });
    expect(wrapper.instance().returnCards()).toEqual([]);

    wrapper.setState({ 
      selected: 'vehicles',
    });
    expect(wrapper.instance().returnCards()).toEqual(['hi', 'bye']);
  })

  it('renderScreen method returns different value based on this.state.selected', () => {
    wrapper.instance().renderScreen()
    expect(wrapper.find('ScrollText').length).toEqual(1)
    expect(wrapper.find('CardContainer').length).toEqual(0)

    wrapper.setState({ 
      navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'people',
      favorites: []
    });

    expect(wrapper.find('ScrollText').length).toEqual(0)
    expect(wrapper.find('CardContainer').length).toEqual(1)
  })

  it('clickFavoriteBtn method adds new object to favorites key in App state', () => {
    let mockObject = { 
      climate: "temperate", 
      created: "2014-12-10T11:35:48.479000Z",
      diameter: "12500",
      edited: "2014-12-20T20:58:18.420000Z",
      films: ["https://swapi.co/api/films/6/", "https://swapi.co/api/films/1/"],
      gravity: "1 standard",
      name: "Alderaan",
      orbital_period: "364",
      population: "2000000000",
      residents: ["https://swapi.co/api/people/5/", "https://swapi.co/api/people/68/", "https://swapi.co/api/people/81/"],
      rotation_period: "24",
      surface_water: "40",
      terrain: "grasslands, mountains",
      url: "https://swapi.co/api/planets/2/"
  }
    wrapper.instance().clickFavoriteButton(mockObject)
    expect(wrapper.state()).toEqual({navFixed: false,
      initialNavPosition: 0,
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      selected: 'films',
      favorites: [mockObject]});
  })
})

// Test allFetchCalls
