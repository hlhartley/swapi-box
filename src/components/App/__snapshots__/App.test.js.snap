// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in correctly 1`] = `
<div
  className="App"
>
  <Header />
  <Navigation
    favorites={Array []}
    initialNavPosition={0}
    navFixed={false}
    people={Array []}
    planets={Array []}
    receiveSelected={[Function]}
    vehicles={Array []}
  />
  <ScrollText
    films={Object {}}
  />
</div>
`;
